Fix "Import Error" sync status with no explanation in the task logs

petercrocker opened the pull request:
This PR addresses issue #6124 by providing clear, actionable error messages when a Git repository is missing its .infrahub.yml configuration file or when the configuration file is invalid.

Previously, when adding a Git repository without a .infrahub.yml file, users would see an "Import Error" sync status with no explanation in the task logs. The only indication was a debug-level log message that wasn't visible to users, making it difficult to diagnose and fix the issue.

Added a new RepositoryConfigurationError exception class that provides clear, user-friendly error messages
Modified get_repository_config() to raise exceptions instead of silently returning None
Error messages now include the repository name, specific issue description, and a link to documentation

Fixes #6124

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **Bug Fixes**
  * Repository configuration now fails fast with clear, actionable errors instead of returning null, preventing silent import/sync inconsistencies.

* **New Features**
  * Added a dedicated repository-configuration error type with a helpful default message for missing or invalid configs.

* **Tests**
  * Added comprehensive tests and fixtures for config loading and error scenarios; updated git-related tests and expectations to align with the new config behavior.

* **Chores**
  * Updated test fixtures to include repository config file for realistic branch/file listings.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

petercrocker committed:
fix https://github.com/opsmill/infrahub/issues/6124

petercrocker committed:
Apply ruff formatting to test files

petercrocker committed:
Remove unnecessary type ignore comments from integrator.py

Since get_repository_config() no longer returns None, the type ignore
comments are no longer needed.

petercrocker committed:
Remove unnecessary type ignore comments from integrator.py

Since get_repository_config() no longer returns None, the type ignore
comments are no longer needed.

petercrocker committed:
Merge branch 'pmc-20251127-infrahubyml-error' of github.com:opsmill/infrahub into pmc-20251127-infrahubyml-error

petercrocker committed:
Apply ruff formatting to import_all_python_files call

petercrocker committed:
Restore type ignore comments required by mypy for Prefect tasks

The type ignore comments are needed because mypy doesn't understand
Prefect's @task decorator return type correctly.

petercrocker committed:
Add .infrahub.yml config file to conflict-01 test fixture

The conflict-01 test fixture was missing a .infrahub.yml configuration
file. With the new RepositoryConfigurationError exception that is now
raised when a repository lacks this config file, the integration tests
using this fixture were failing.

This adds a minimal .infrahub.yml to fix the integration tests:
- test_run_pipeline_validate_requested_jobs
- test_run_generators_validate_requested_jobs

petercrocker committed:
Merge branch 'develop' into pmc-20251127-infrahubyml-error

petercrocker committed:
Merge remote-tracking branch 'origin/develop' into pmc-20251127-infrahubyml-error

# Conflicts:
#	backend/infrahub/git/integrator.py
#	backend/tests/unit/git/test_git_repository.py

petercrocker committed:
ruff

petercrocker committed:
Merge remote-tracking branch 'origin/develop' into pmc-20251127-infrahubyml-error

# Conflicts:
#	backend/tests/component/git/test_repository_config.py

petercrocker committed:
Merge branch 'develop' into pmc-20251127-infrahubyml-error

ajtmccarty commented on backend/infrahub/exceptions.py:
```diff
@@ -116,6 +116,23 @@ def __init__(
+
+    This exception provides clear error messages when a Git repository
+    lacks a .infrahub.yml/.infrahub.yaml file or when the file cannot
+    be parsed correctly.
```
I'd say this can be removed. it adds nothing that would not be covered by `message`

ajtmccarty commented on backend/infrahub/exceptions.py:
```diff
@@ -116,6 +116,23 @@ def __init__(
+            identifier=identifier,
+            message=message
+            or f"Repository '{identifier}' is missing a configuration file. "
+            f"Please add a '.infrahub.yml' or '.infrahub.yaml' file to the repository root.",
```
the default should be more generic. if the error is actually raised b/c of a missing `.infrahub.y[a]ml`, then the error message will say so

ajtmccarty commented on backend/tests/component/git/test_git_read_only_repository.py:
```diff
@@ -82,7 +82,9 @@ async def test_sync_from_remote_new_ref(git_repo_01_read_only: InfrahubReadOnlyR
-    await repo.sync_from_remote()
+    # Mock import_objects_from_files since we're testing git sync, not import functionality
+    with patch("infrahub.git.repository.InfrahubReadOnlyRepository.import_objects_from_files", new_callable=AsyncMock):
+        await repo.sync_from_remote()
```
if this is actually necessary, I think it can just be `repo.import_objects_from_files = AsyncMock()` which is simpler than a patch

ajtmccarty commented on backend/tests/component/git/test_git_repository.py:
```diff
@@ -499,20 +499,18 @@ async def test_sync_new_branch(
-    await repo.sync()
+    # Mock import_objects_from_files since we're testing git sync, not import functionality
+    with patch("infrahub.git.repository.InfrahubRepository.import_objects_from_files", new_callable=AsyncMock):
+        await repo.sync()
```
samesies down here for `repo.import_objects_from_files = AsyncMock()`.
not sure if it is necessary, but usually we want to assert that a mock is called with the expected arguments. this would have been previously handled by the removed `httpx_mock`

gmazoyer commented on backend/tests/component/git/test_repository_config.py:
```diff
@@ -0,0 +1,199 @@
+        self, git_upstream_repo_01: dict[str, str | Path], git_repos_dir: Path
+    ) -> InfrahubRepository:
+        """Create a repository with an invalid YAML config file."""
+        from git import Repo as GitRepo
```
Ideally import should be in the import block at the top of the file.

petercrocker committed:
refactor: address code review feedback

- Make RepositoryConfigurationError default message generic
- Move git.Repo import to top of test files
- Simplify test mocks by replacing patch context managers with direct assignment
- Add assertions to verify mocked methods are called
- Fix pre-existing type errors with client.schema checks

petercrocker committed:
test: address CodeRabbit review suggestions

- Add assertion to verify docs.infrahub.app URL is present in error messages
- Reorganize test fixtures to improve readability by moving repo_with_valid_config
  fixture to be with other fixtures at the top of the test class

petercrocker committed:
test: fix test isolation by cloning repos per-test

Previously, test fixtures modified the shared git_upstream_repo_01 repository
directly, causing commits to leak between tests. This could lead to
unpredictable test behavior and failures.

Changes:
- repo_with_invalid_yaml: Clone upstream repo to isolated directory
- repo_with_invalid_format: Clone upstream repo to isolated directory
- repo_with_valid_config: Clone upstream repo to isolated directory

Each fixture now:
1. Clones git_upstream_repo_01 to a unique path under git_repos_dir
2. Performs config file modifications on the clone
3. Creates InfrahubRepository pointing to the clone

This ensures complete test isolation and prevents state leakage.

petercrocker committed:
docs: align fixture docstrings with Google-style format

Updated all test fixture docstrings to follow Google-style format as per
coding guidelines. Each docstring now includes:

- Args section describing the fixture parameters
- Returns section describing the returned value

Fixtures updated:
- repo_without_config: Added Args/Returns sections
- repo_with_invalid_yaml: Added Args/Returns sections
- repo_with_invalid_format: Added Args/Returns sections
- repo_with_valid_config: Added Args/Returns sections

This ensures consistency with project coding standards and improves
documentation clarity.

petercrocker committed:
style: fix docstring formatting in exceptions.py

Ruff formatting fix for RepositoryConfigurationError docstring.

petercrocker committed:
Fix mock approach for Pydantic models in repository tests

The simplified direct assignment approach (repo.import_objects_from_files = AsyncMock())
doesn't work with Pydantic models because they have strict field validation and don't
allow arbitrary attribute assignment.

Changed to use patch() with full module path instead of direct assignment or patch.object(),
which allows mocking methods on Pydantic model instances without triggering validation errors.

petercrocker committed:
refactor: replace deprecated timezone.utc with UTC

Use the newer UTC constant instead of timezone.utc for better
compatibility with Python 3.11+.

petercrocker committed:
Revert "refactor: replace deprecated timezone.utc with UTC"

This reverts commit d7c90496bf77f069a27d870f3bf4d2378d2e9abb.

petercrocker committed:
Merge branch 'develop' into pmc-20251127-infrahubyml-error

petercrocker committed:
fix: resolve CI test failures for repository import tests

- Update infrahub-test-fixture-01 tarball to include .infrahub.yml config
  file in all branches (main, branch01, branch02, clean-branch). This is
  required by the new RepositoryConfigurationError validation.

- Add default_permission_backend fixture to test_trigger_repository_import
  and test_import_read_only_repository_last_commit tests. This initializes
  the permission backend so admin permissions are properly loaded during
  GraphQL mutation execution.

petercrocker committed:
fix: update tests to use new commit hashes from updated fixture

The infrahub-test-fixture-01 tarball was updated to include .infrahub.yml
configuration files, which changed the commit hashes. This updates:

- test_git_repository.py: Update hardcoded commit hashes for main, branch01,
  and branch02 to match the new fixture
- test_git_read_only_repository.py: Update hardcoded commit hashes for
  branch01 and branch02
- test_repository_config.py: Update repo_without_config fixture to clone
  the repo and remove the config file, since the fixture now includes one

petercrocker commented:
@coderabbitai review

petercrocker committed:
fix: rebuild test fixture with proper commit history to avoid merge conflicts

The previous fixture had .infrahub.yml added to each branch separately,
creating diverging commit histories that caused merge conflicts during
git operations. This rebuild uses git filter-branch to add .infrahub.yml
to ALL commits, preserving the original branching structure where all
branches inherit the file from their common ancestor.

Updated all hardcoded commit hashes in tests to match the new fixture.

petercrocker committed:
fix: update test assertions for .infrahub.yml in fixture

- Update file count assertions in test_find_files and test_find_files_by_commit
  to account for .infrahub.yml now being present in the test fixture
- Add .infrahub.yml to expected file lists in test_list_all_files
- Mock import_objects_from_files in test_import_read_only_repository_last_commit
  to prevent AttributeError when accessing client.schema

petercrocker committed:
fix: correct file count assertions for duplicate .infrahub.yml match

The glob patterns **/*.yml and **/.*.yml both match .infrahub.yml,
causing it to be counted twice in the find_files results. Updated
assertions to reflect the correct counts (4 yml files, 6 yml+j2 files).

petercrocker committed:
Merge branch 'develop' into pmc-20251127-infrahubyml-error

petercrocker commented:
@coderabbitai review

petercrocker committed:
refactor: simplify mock by assigning directly to instance method

Per code review feedback, use repo.import_objects_from_files = AsyncMock()
instead of patch() for test_sync_from_remote_new_ref since we have access
to the repo instance.

petercrocker committed:
fix: use patch.object for Pydantic model method mocking

Direct attribute assignment doesn't work on Pydantic models since they
validate field assignments. Use patch.object() instead which properly
handles method mocking on Pydantic instances.

petercrocker committed:
fix: revert to class-level patch for Pydantic model compatibility

patch.object doesn't work with Pydantic models because Pydantic intercepts
both __setattr__ and __delattr__. The class-level patch is required to
avoid triggering Pydantic's attribute validation.

ajtmccarty committed:
clean up tests

ajtmccarty reviewed (approved):
I don't really understand what is going on in the updated `test_git[_read_only]_repository` tests, but they seem to be working
I pushed a commit to clean up a few small things in the tests instead of asking you to do it

petercrocker commented:
> I don't really understand what is going on in the updated `test_git[_read_only]_repository` tests, but they seem to be working I pushed a commit to clean up a few small things in the tests instead of asking you to do it

The read-only repo test is loading a tarball of a valid git repo. Since it's now testing if a `.infrahub.yml` file exists in a repo, I had to create a new tarball, and update the hashes everywhere to that tarball. Only difference is adding a `.infrahub.yml` file to that test.
